aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -542,6 +542,91 @@ Tensor &nuclear_norm_out(Tensor& result, const Tensor& self, bool keepdim) {
   return at::sum_out(result, std::get<1>(at::svd(self)), 0, keepdim);
 }
 
+// Return a permutation with the given axes moved to the end.
+static Tensor move_to_end(const Tensor& self, IntArrayRef axes) {
+  const std::vector<int64_t> a = axes.vec();
+  std::vector<int64_t> perm;
+
+  for (int64_t i = 0; i < self.ndimension(); i++) {
+    auto it = std::find(a.begin(), a.end(), i);
+    if (it == a.end()) {
+       perm.push_back(i);
+    }
+  }
+  for (auto i : a) {
+    perm.push_back(i);
+  }
+
+  return self.permute(perm);
+}
+
+// Non-optimized batched svd implementation. This can be merged with at::svd
+// once at::svd has been ported to ATen.
+static std::tuple<Tensor, Tensor, Tensor>
+_batch_svd(const Tensor& self, bool some, bool compute_uv)
+{
+  const int64_t ndim = self.ndimension();
+
+  TORCH_CHECK(
+      ndim >= 2,
+      "Expected a tensor with at least 2 dimensions, but got a ",
+      ndim,
+      " dimensions tensor instead.");
+
+  if (ndim == 2) {
+    return at::svd(self, some, compute_uv);
+  }
+
+  const int64_t n = self.size(-2);
+  const int64_t m = self.size(-1);
+  const int64_t k = std::min<int64_t>(n, m);
+  const int64_t nn = (some && compute_uv) ? k : n;
+  const int64_t mm = (some && compute_uv) ? k : m;
+  const int64_t p = batchCount(self);
+
+  Tensor t = self.reshape({p, n, m}).contiguous();
+  Tensor u = at::zeros({p, n, nn}, self.options());
+  Tensor s = at::zeros({p, k}, self.options());
+  Tensor v = at::zeros({p, m, mm}, self.options());
+
+  for (int64_t i = 0; i < p; i++) {
+    auto tuple = at::svd(t[i], some, compute_uv);
```

Comment:
@ezyang Could you help out here?  We were wondering if it is guaranteed that `t[i]` works on a non-contiguous batch dimension.  If it does, we could avoid one call to `contiguous()`.

I ran some tests, and it seems to work in practice, but it is almost too good to be true. :)

Reply:
I think it does, since this operation is effectively calls `at::select` internally.

Reply:
Good point, thanks for the reference!  I'll update the PR.